RavenDB-15976 Rate flow control for indexes journal writes.

redknightlois opened the pull request:
Concurrent writes to the indexes now can be synchronized at the database or server level. The functionality is implemented at the Voron level so while now the only synchronizing writes happen when indexing and can be controlled through configuration:
Indexing.JournalSynchronizationMode = {Immediate|Database|Server} = Database
Indexing.JournalMaxConcurrentWrites = 4
Indexing.JournalMaxConcurrentWritesSizeInMb = 128
This same mechanism can be used for any other purpose at the Journal level.

redknightlois committed:
RavenDB-15976 Rate flow control for indexes journal writes.

Concurrent writes to the indexes now can be synchronized at the database or server level. The functionality is implemented at the Voron level so while now the only synchronizing writes happen when indexing and can be controlled through configuration:
Indexing.JournalSynchronizationMode = {Immediate|Database|Server} = Database
Indexing.JournalMaxConcurrentWrites = 4
Indexing.JournalMaxConcurrentWritesSizeInMb = 128
This same mechanism can be used for any other purpose at the Journal level.

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -222,6 +223,25 @@ public virtual bool RunInMemory
+        [Description("Synchronization mode for the indexes")]
+        [DefaultValue(IndexWriteSynchronizationMode.Database)]
+        [IndexUpdateType(IndexUpdateType.None)]
+        [ConfigurationEntry("Indexing.JournalSynchronizationMode", ConfigurationEntryScope.ServerWideOnly)]
```
Any reason we shouldn't make it a per db configuration?

redknightlois replied:
We could. But it makes a lot more sense to do that at the server level as that would give better behavior at the server level. I would say that unless we find the case where it really matters to control a single database write synchronization I would leave it Server wide.

ayende replied:
`ServerWideOnly` vs `ServerWideOrDatabase` still allows you to define it globally.
This matters, because db only configuration is something that you can do in the cloud without needing manual assitance.

ppekrol replied:
I also think it should be `ServerWideOrDatabase`

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -222,6 +223,25 @@ public virtual bool RunInMemory
         [ConfigurationEntry("Indexing.NuGetPackageSourceUrl", ConfigurationEntryScope.ServerWideOnly)]
         public string NuGetPackageSourceUrl { get; set; }
 
+        [Description("Synchronization mode for the indexes")]
```
```suggestion
        [Description("Determine how RavenDB should coordinate expensive journal writes for indexes. This is used to reduce I/O contention with many indexes.")]
```

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -168,6 +171,20 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
                 _hasClusterTransaction = new AsyncManualResetEvent(DatabaseShutdown);
                 IdentityPartsSeparator = '/';
+
+                this.IndexWritesSynchronization = new StorageEnvironmentSynchronization();
```
Move this to the `case`'s `default`?

redknightlois replied:
Yes, I am a bit hesitant if `switch` `default` because they tend to mask errors, but in this case which is essentially an ignore, there is no reason why not.

ayende commented on src/Raven.Server/Indexing/IndexWriteSynchronizationMode.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public enum IndexWriteSynchronizationMode
+    {
+        Immediate,
```
```suggestion
        None,
```

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,49 +28,87 @@ public unsafe class JournalWriter : IJournalWriter
+
+            // The write synchronization mechanism will require a max size and to be aligned to 4kb.  
+            _writeSynchronization = writeSynchronization;
+            _maxNumberOf4KbInSingleWrite = writeSynchronization.MaxWriteSize / 4096;
```
```suggestion
            _maxNumberOf4KbInSingleWrite = writeSynchronization.MaxWriteSize /  (4L * Constants.Size.Kilobyte);
```

Let's keep this consistent.

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,49 +28,87 @@ public unsafe class JournalWriter : IJournalWriter
+                }
 
-                if (error == ERROR_WORKING_SET_QUOTA && _log.IsOperationsEnabled && _workingSetQuotaLogged == false)
+                // Then we will write the last chunk. 
```
We can remove this.

redknightlois replied:
I suffered a bit after coming back the lack of comments, agree this particular one is not so important, but I will be writing a lot more from now on.

ayende replied:
That isn't about the comments, that is about handling the last portion, there isn't going to be one.
It is always 4KB aligned.

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,49 +28,87 @@ public unsafe class JournalWriter : IJournalWriter
-                    PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path: {FileName.FullPath} Size: {numberOf4Kb * 4L * Constants.Size.Kilobyte}, numberOf4Kb={numberOf4Kb}");
+                // We will separate them in chunks of max size. 
+                long iterations = numberOf4Kb / _maxNumberOf4KbInSingleWrite;
+                long leftovers = numberOf4Kb % _maxNumberOf4KbInSingleWrite;
```
You can assert that this `leftovers` is zero here, we should always be passing 4KB aligned buffers here.

redknightlois replied:
Not really, those are 4kb aligned blocks :) a zero on both means no write in effect.

ayende replied:
The key here is that we want to remove the code to handle the left overs

redknightlois replied:
Leftovers are 4kb aligned blocks that need to be written. Unless what you mean is that we could simplify that code instead of using 2 sequences, use a single that handle both cases.

ayende commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,49 @@ namespace Voron
+            this.MaxWriteSize = maxWriteSize;
+        }        
+
+        public SynchronizationSection Enter()
```
We are going to need to plug this into the I/O metrics, so we can see how long we waited for this.

ayende commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,49 @@ namespace Voron
+            public SynchronizationSection(Semaphore context)
+            {
+                this._context = context;
+                if (this._context != null)
```
Under what scenario this can be null and we want to accept that?

redknightlois replied:
`None` is the scenario. We don't need to pay the synchronization cost if it is not used.

ayende commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,49 @@ namespace Voron
+            {
+                this._context = context;
+                if (this._context != null)
+                    this._context.WaitOne();
```
We need to make sure that this is an _interruptible_ wait.
Right now, if the index is waiting for a write, we can't shut down the server, for example.

redknightlois replied:
Makes sense, where do we get that interruption callback from?

ayende replied:
The easiest is to the use the `DatabaseShutdown` cancellation token. IIRC, there is one for each indexc.

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,49 +28,87 @@ public unsafe class JournalWriter : IJournalWriter
+                byte* currentBufferPtr = buffer;
+                for (long i = 0; i < iterations; i++ )
+                {
+                    using (_writeSynchronization.Enter())
```
I worry about fairness here. The semaphore isn't fair, which means that a tight _release_ and _acquire_ is going to see the same index keeping the lock for a long time.
That isn't necessarily a bad idea, mind, because we'll complete this one faster and there is a limit to how long we can stall here, but need to consider this explicitly.

redknightlois replied:
I thought exactly about that, performance-wise hitting the same index/file is always faster than jumping around on sector addressable media. Therefore we kinda want it to be a bit unfair if possible.

ayende replied:
Need a comment explaining that, I think.

ayende reviewed (commented):
We need to see what kind of performance difference this makes.
The scenarios in question should be:

* One db - one index
* One db - 10 indexes
* 5 dbs, - 1 index each
* 5 dbs - 10 indexes each

And then we want to measure:
* Amount of IOPS generated
* Time for the index to complete.

redknightlois committed:
RavenDB-15976 Addressing review comments.

redknightlois committed:
RavenDB-15976 Ensure we honor database and server shutdown

redknightlois committed:
RavenDB-15976 Changed to report the wait time instead of the file access to the type.

ppekrol commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -222,6 +223,25 @@ public virtual bool RunInMemory
+        [DefaultValue(128)]
+        [IndexUpdateType(IndexUpdateType.None)]
+        [ConfigurationEntry("Indexing.JournalMaxConcurrentWritesSizeInMb", ConfigurationEntryScope.ServerWideOnly)]
+        public int JournalMaxConcurrentWritesSizeInMegabytes { get; set; }        
```
Use `Size` type e.g.
https://github.com/ravendb/ravendb/blob/785eb6a41c1faf45331cfe17ebe8a216d0845c6a/src/Raven.Server/Config/Categories/StorageConfiguration.cs#L54-L58

ppekrol replied:
The property then should be `JournalMaxConcurrentWritesSize` and the configuration entry `Indexing.JournalMaxConcurrentWritesSizeInMb`

ppekrol commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -96,6 +97,8 @@ public void ResetIdleTime()
             _lastIdleTicks = DateTime.MinValue.Ticks;
         }
 
+        public StorageEnvironmentSynchronization IndexWritesSynchronization { get; private set; }
```
This should go into IndexStore

ppekrol commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -177,7 +179,13 @@ public ServerStore(RavenConfiguration configuration, RavenServer server)
-                           uri.Port != 0;
+                           uri.Port != 0;                       
+
+            this.IndexWritesSynchronization = new StorageEnvironmentSynchronization(
```
remove `this`

ppekrol commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -168,6 +171,22 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
+                    case IndexWriteSynchronizationMode.Database:
+                        this.IndexWritesSynchronization = new StorageEnvironmentSynchronization(
+                                  serverStore.Configuration.Indexing.JournalMaxConcurrentWrites,
+                                  serverStore.Configuration.Indexing.JournalMaxConcurrentWritesSizeInMegabytes * Sparrow.Global.Constants.Size.Megabyte,
```
use `Size` type as a parameter, we do not want to deal with conversions manually. You will get it from the configuration after applying my previous remark.

ppekrol commented on src/Sparrow.Server/Meters/IoMetrics.cs:
```diff
@@ -10,12 +10,12 @@ public enum MeterType
         {
             Compression,
             JournalWrite,
+            JournalWait,
```
We need to check if this requires any adjustments in Studio IO Metrics graph

redknightlois replied:
Yes, who would be the one to hand over this ticket when this is done, or do I create a new ticket?

ppekrol replied:
Assign to @arekpalinski he will distrubte

ppekrol commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,49 +28,96 @@ public unsafe class JournalWriter : IJournalWriter
+                    writeMetrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWait, _maxNumberOf4KbBlocks * 4L * Constants.Size.Kilobyte);
+                    using (_writeSynchronization.Enter())
+                    {
+                        writeMetrics.SetFileSize(_maxNumberOf4KbBlocks * (4L * Constants.Size.Kilobyte));
```
is this correct? each iteration is writing less?

redknightlois replied:
No `_maxNumberOf4KbBlocks` comes from the constructor, it won't change. The first iteration will write always the max number of blocks possible through configuration. The only case that is not the case is when `leftovers` is non-zero. The only 2 conditions when it might happen is smaller than biggest block write or when we processed everything and we are left with a partial block at the end.

ppekrol commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,54 @@ namespace Voron
+
+            public SynchronizationSection(SemaphoreSlim context, CancellationToken cancellationToken)
+            {
+                this._context = context;
```
in our code styling we do not use `this`

ppekrol commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,54 @@ namespace Voron
+        }
+
+        private readonly SemaphoreSlim _writeSemaphore;
+        public readonly long MaxWriteSize;
```
rename to `MaxWriteSizeInBytes` ? Maybe leave it as `Size` ?

redknightlois replied:
Like `MaxWriteSizeInBytes` more, because it is more explicit.

ppekrol replied:
https://github.com/ravendb/ravendb/blob/v5.1/src/Sparrow/Size.cs

redknightlois replied:
Ahhhh you mean to use the `Size` type... I wasn't understanding... roger that

redknightlois committed:
RavenDB-15976 Review changes.

ml054 committed:
RavenDB-15976 Rate flow control for indexes journal writes.

Danielle9897 committed:
RavenDB-15976 Fix JournalWait envPath

redknightlois committed:
RavenDB-15976 Simplified code for journal write

redknightlois committed:
RavenDB-15976 Cleanup of Journal code.

The `IJournalWriter` interface is a remnant of the older RavenDB 3.0 design where ESENT and Voron were present. It serves no purpose now.

redknightlois committed:
RavenDB-15976 Writes to the index now are shown in pieces

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -222,6 +223,26 @@ public virtual bool RunInMemory
+        [Description("Determine how RavenDB should coordinate expensive journal writes for indexes. This is used to reduce I/O contention with many indexes.")]
+        [DefaultValue(IndexWriteSynchronizationMode.Database)]
+        [IndexUpdateType(IndexUpdateType.None)]
+        [ConfigurationEntry("Indexing.JournalSynchronizationMode", ConfigurationEntryScope.ServerWideOnly)]
```
```suggestion
        [ConfigurationEntry("Indexing.JournalSynchronizationMode", ConfigurationEntryScope.ServerWideOrDatabase)]
```

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -222,6 +223,26 @@ public virtual bool RunInMemory
+        public IndexWriteSynchronizationMode JournalSynchronizationMode { get; set; }
+
+        [Description("Maximum allowed concurrent writes for the indexes")]
+        [DefaultValue(4)]
```
Isn't that a high value, shouldn't we favor 1 / 2 ?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -84,6 +88,22 @@ public IndexStore(DocumentDatabase documentDatabase, ServerStore serverStore)
+            switch (_serverStore.Configuration.Indexing.JournalSynchronizationMode)
+            {
+                case IndexWriteSynchronizationMode.Server:
+                    IndexWritesSynchronization = serverStore.IndexWritesSynchronization;
```
Shouldn't we create a new instance here that uses `DatabaseShutdown`, instead? But the same semaphore as the server?

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,50 +29,75 @@ public unsafe class JournalWriter : IJournalWriter
             FileName = filename;
             _log = LoggingSource.Instance.GetLogger<JournalWriter>(options.BasePath.FullPath);
 
+            if (writeSynchronization == null)
+                writeSynchronization = new StorageEnvironmentSynchronization();
+
+            // The write synchronization mechanism will require a max size and to be aligned to 4kb.  
+            _writeSynchronization = writeSynchronization;
```
```suggestion
            // The write synchronization mechanism will require a max size and to be aligned to 4kb.  
            _writeSynchronization = writeSynchronization ?? new StorageEnvironmentSynchronization();
```

ayende commented on src/Voron/Impl/Journal/JournalWriter.cs:
```diff
@@ -28,50 +29,75 @@ public unsafe class JournalWriter : IJournalWriter
+                using (_writeSynchronization.Enter())
+                {
+                    writeMetrics.SetFileSize(blocksToWrite * (4L * Constants.Size.Kilobyte));
+                    writeMetrics.Dispose();
```
What happens on error? Can we wrap this in a `using` ?

ayende commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -38,8 +38,58 @@ namespace Voron
+                if (_context != null)
+                {
+                    _context.Wait(cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
```
Shouldn't we always call this?
